src/System.Xml.XmlSerializer/src/System/Xml/Serialization/Compilation.cs

```diff
@@ -305,21 +303,27 @@ public override int GetHashCode()
 
     internal class TempAssemblyCache
     {
-        private Hashtable _cache = new Hashtable();
+        private Dictionary<TempAssemblyCacheKey, TempAssembly> _cache = new Dictionary<TempAssemblyCacheKey, TempAssembly>();
```

Comment:
Hmmm, be careful here.  It looks like this class might be relying on a behavior of ```Hashtable``` that ```Dictionary<TKey, TValue>``` doesn't have: it's safe to use ```Hashtable``` concurrently with many readers and a single writer, whereas it's not safe to be concurrently reading and writing a dictionary.  I mention this because the Add method is using a lock, but reading in the indexer isn't: that implies that the author of this code was serializing the writers to ensure that there's only one at a time, allowing this to be safe with Hashtable, but now even with that lock in Add, if code concurrently uses the indexer and the Add method, that could cause problems for the readers (it won't corrupt the dictionary, but it could cause the readers to experience unexpected exceptions, get back unexpected results, etc.)